### ISSUE TITLE
allow INSERT statement with a query

### ISSUE THREAD

**Issue by billmaimone** [mapd/mapd-core#109]
Add support for INSERT INTO table  ( col..) SELECT ...
Care must be taken to ensure that datatypes from the query match up with the insert table.
Current workaround for this is COPY TO/FROM, but avoiding intermediate the transfer and file write would be faster and more convenient.

**Comment by fleapapa** [comment]
Related with this issue, i'm curious how mapd_server handles a SELECT that returns data of a size larger than system memory. On my VM with 14GB free memory, query "SELECT * from trips" failed due to OOM. Also failed is another query "copy (select * from trips) to '/mnt/tmp/trips.csv'", with similar INFO logs as below shows. The table _trips_ has 512M rows, using approximately 47GB disk space. 

Is there any query option that can work around this limitation? If no, does it mean that, regardless intermediate file write, a COPY TO query of data larger than available memory will fail?

If that is the case, what is the complexity of a change to current design to make the part of code that returns query result to work in a **streaming** manner? 

I am curious about a parameter _use_resume_param_ in function NativeCodegen.cpp:_bind_pos_placeholders_ and what it tries to "resume". Somehow with gdb i can't trace into  **RuntimeFunctions.bc** to learn the effect of the parameter and how related LLVM code works. Any advice in this regard will be appreciated.

```
E1017 20:09:34.659145  8212 MapDHandler.cpp:139] This build isn't CUDA enabled, will run on CPU
I1017 20:09:34.671910  8212 DataMgr.cpp:120] cpuSlabSize is 4096M
I1017 20:09:34.698573  8212 FileMgr.cpp:173] Completed Reading table's file metadata, Elasped time : 11ms Epoch: 0 files read: 0 table location: '/mnt/trip/mapd/data/mapd_data/table_0_0/'
I1017 20:09:34.707365  8212 Calcite.cpp:156] Creating Calcite Handler,  Calcite Port is 9093 base data dir is /mnt/trip/mapd/data
I1017 20:09:34.707384  8212 Calcite.cpp:95] Running calcite server as a daemon
I1017 20:09:35.374925  8212 Calcite.cpp:124] Calcite server start took 600 ms 
I1017 20:09:35.374971  8212 Calcite.cpp:125] ping took 18 ms 
I1017 20:09:35.380465  8212 Calcite.cpp:281] [{"name":"Tan","ret":"double","args":["double"]},{"name":"Truncate__","ret":"float","args":["float","i32"]},{"name":"ln","ret":"double","args":["double"]},{"name":"distance_in_meters__","ret":"double","args":["float","float","float","float"]},{"name":"Floor__","ret":"float","args":["float"]},{"name":"Floor__2","ret":"i32","args":["i32"]},{"name":"Floor__3","ret":"i64","args":["i64"]},{"name":"Truncate","ret":"double","args":["double","i32"]},{"name":"Floor__1","ret":"i16","args":["i16"]},{"name":"radians","ret":"double","args":["double"]},{"name":"degrees","ret":"double","args":["double"]},{"name":"Ceil__","ret":"float","args":["float"]},{"name":"ln__","ret":"double","args":["float"]},{"name":"approx_distance_in_meters","ret":"double","args":["float","float","float","float"]},{"name":"Ceil__2","ret":"i32","args":["i32"]},{"name":"Ceil__1","ret":"i16","args":["i16"]},{"name":"Ceil__3","ret":"i64","args":["i64"]},{"name":"power","ret":"double","args":["double","double"]},{"name":"conv_4326_900913_x","ret":"double","args":["double"]},{"name":"conv_4326_900913_y","ret":"double","args":["double"]},{"name":"Atan","ret":"double","args":["double"]},{"name":"Floor","ret":"double","args":["double"]},{"name":"Log10","ret":"double","args":["double"]},{"name":"Truncate__1","ret":"i16","args":["i16","i32"]},{"name":"Truncate__2","ret":"i32","args":["i32","i32"]},{"name":"Log","ret":"double","args":["double"]},{"name":"Truncate__3","ret":"i64","args":["i64","i32"]},{"name":"Cos","ret":"double","args":["double"]},{"name":"Log__","ret":"double","args":["float"]},{"name":"Log10__","ret":"double","args":["float"]},{"name":"Asin","ret":"double","args":["double"]},{"name":"Cot","ret":"double","args":["double"]},{"name":"Tan__","ret":"double","args":["float"]},{"name":"distance_in_meters","ret":"double","args":["double","double","double","double"]},{"name":"round_to_digit","ret":"double","args":["double","i32"]},{"name":"Atan2","ret":"double","args":["double","double"]},{"name":"Acos","ret":"double","args":["double"]},{"name":"pi","ret":"double","args":[]},{"name":"Sin","ret":"double","args":["double"]},{"name":"Ceil","ret":"double","args":["double"]},{"name":"Exp","ret":"double","args":["double"]}]
E1017 20:09:35.403663  8212 MapDHandler.cpp:171] No GPUs detected, falling back to CPU mode
I1017 20:09:35.403697  8212 MapDHandler.cpp:180] Started in CPU mode
I1017 20:10:05.424568  8275 MapDHandler.cpp:311] User mapd connected to database mapd
I1017 20:10:05.472877  8275 MapDHandler.cpp:2206] copy (select * from trips) to '/mnt/tmp/trips.csv';
I1017 20:10:05.477603  8275 MapDHandler.cpp:2234] passing query to legacy processor
I1017 20:10:05.489598  8275 Calcite.cpp:247] User mapd catalog mapd sql 'select * from trips;'
I1017 20:13:25.988349  8278 FileMgr.cpp:173] Completed Reading table's file metadata, Elasped time : 199575ms Epoch: 125234 files read: 506 table location: '/mnt/trip/mapd/data/mapd_data/table_1_1/'
I1017 20:13:38.456360  8278 Catalog.cpp:1511] Instantiating Fragmenter for table trips took 212085ms
I1017 20:13:38.626971  8278 Catalog.cpp:1557] Time to load Dictionary 1 was 170ms
I1017 20:13:38.763283  8278 Catalog.cpp:1557] Time to load Dictionary 2 was 136ms
I1017 20:13:38.881168  8278 Catalog.cpp:1557] Time to load Dictionary 3 was 117ms
I1017 20:13:38.990425  8278 Catalog.cpp:1557] Time to load Dictionary 4 was 109ms
I1017 20:13:39.280971  8275 Calcite.cpp:260] Time in Thrift 13 (ms), Time in Java Calcite server 213778 (ms)
I1017 20:13:39.716213 10892 BufferMgr.cpp:266] ALLOCATION slab of 8388608 pages (4294967296B) created in 0 ms CPU_MGR:0
I1017 20:14:30.122083 10893 BufferMgr.cpp:266] ALLOCATION slab of 8388608 pages (4294967296B) created in 0 ms CPU_MGR:0
I1017 20:15:30.326671 10895 BufferMgr.cpp:266] ALLOCATION slab of 6633401 pages (3396301312B) created in 0 ms CPU_MGR:0
E1017 20:16:21.526213 10898 BufferMgr.cpp:370] ALLOCATION failed to find 300000256B throwing out of memory CPU_MGR:0
E1017 20:16:21.682365 10899 BufferMgr.cpp:370] ALLOCATION failed to find 503252992B throwing out of memory CPU_MGR:0
E1017 20:16:21.683790 10894 BufferMgr.cpp:370] ALLOCATION failed to find 300000256B throwing out of memory CPU_MGR:0
E1017 20:16:21.685151 10893 BufferMgr.cpp:370] ALLOCATION failed to find 600000000B throwing out of memory CPU_MGR:0
E1017 20:16:21.685667 10892 BufferMgr.cpp:370] ALLOCATION failed to find 600000000B throwing out of memory CPU_MGR:0
E1017 20:16:21.686103 10895 BufferMgr.cpp:370] ALLOCATION failed to find 300000256B throwing out of memory CPU_MGR:0
I1017 20:16:24.503365 10897 BufferMgr.cpp:375] ALLOCATION failed to find 300000256B free. Forcing Eviction. Eviction start 0 Number pages requested 585938 Best Eviction Start Slab 0 CPU_MGR:0
I1017 20:16:29.814640 10897 BufferMgr.cpp:375] ALLOCATION failed to find 600000000B free. Forcing Eviction. Eviction start 585938 Number pages requested 1171875 Best Eviction Start Slab 0 CPU_MGR:0
I1017 20:16:40.076355 10897 BufferMgr.cpp:375] ALLOCATION failed to find 600000000B free. Forcing Eviction. Eviction start 3222659 Number pages requested 1171875 Best Eviction Start Slab 0 CPU_MGR:0
I1017 20:16:52.534536 10897 BufferMgr.cpp:375] ALLOCATION failed to find 300000256B free. Forcing Eviction. Eviction start 1757813 Number pages requested 585938 Best Eviction Start Slab 0 CPU_MGR:0
I1017 20:16:57.187934 10897 BufferMgr.cpp:375] ALLOCATION failed to find 600000000B free. Forcing Eviction. Eviction start 4687503 Number pages requested 1171875 Best Eviction Start Slab 0 CPU_MGR:0
I1017 20:17:06.715833 10897 BufferMgr.cpp:375] ALLOCATION failed to find 600000000B free. Forcing Eviction. Eviction start 5859378 Number pages requested 1171875 Best Eviction Start Slab 0 CPU_MGR:0
I1017 20:17:17.180187 10897 BufferMgr.cpp:375] ALLOCATION failed to find 600000000B free. Forcing Eviction. Eviction start 0 Number pages requested 1171875 Best Eviction Start Slab 1 CPU_MGR:0
I1017 20:17:27.879951 10897 BufferMgr.cpp:375] ALLOCATION failed to find 600000000B free. Forcing Eviction. Eviction start 1171875 Number pages requested 1171875 Best Eviction Start Slab 1 CPU_MGR:0
I1017 20:17:38.307312 10897 BufferMgr.cpp:375] ALLOCATION failed to find 600000000B free. Forcing Eviction. Eviction start 2343750 Number pages requested 1171875 Best Eviction Start Slab 1 CPU_MGR:0
I1017 20:17:49.696372 11245 BufferMgr.cpp:375] ALLOCATION failed to find 300000256B free. Forcing Eviction. Eviction start 2343751 Number pages requested 585938 Best Eviction Start Slab 0 CPU_MGR:0
I1017 20:17:54.284348 11245 BufferMgr.cpp:375] ALLOCATION failed to find 300000256B free. Forcing Eviction. Eviction start 7031253 Number pages requested 585938 Best Eviction Start Slab 0 CPU_MGR:0
I1017 20:17:58.744581 11247 BufferMgr.cpp:375] ALLOCATION failed to find 300000256B free. Forcing Eviction. Eviction start 7617191 Number pages requested 585938 Best Eviction Start Slab 0 CPU_MGR:0
I1017 20:18:04.263002 11248 BufferMgr.cpp:375] ALLOCATION failed to find 300000256B free. Forcing Eviction. Eviction start 3515625 Number pages requested 585938 Best Eviction Start Slab 1 CPU_MGR:0
I1017 20:18:09.654181 11249 BufferMgr.cpp:375] ALLOCATION failed to find 300000256B free. Forcing Eviction. Eviction start 4101563 Number pages requested 585938 Best Eviction Start Slab 1 CPU_MGR:0
I1017 20:18:14.792775 11250 BufferMgr.cpp:375] ALLOCATION failed to find 300000256B free. Forcing Eviction. Eviction start 4687501 Number pages requested 585938 Best Eviction Start Slab 1 CPU_MGR:0
I1017 20:18:25.347380 11246 BufferMgr.cpp:375] ALLOCATION failed to find 300000256B free. Forcing Eviction. Eviction start 5859377 Number pages requested 585938 Best Eviction Start Slab 1 CPU_MGR:0
I1017 20:18:34.596098 11247 BufferMgr.cpp:375] ALLOCATION failed to find 300000256B free. Forcing Eviction. Eviction start 7031253 Number pages requested 585938 Best Eviction Start Slab 1 CPU_MGR:0
I1017 20:18:41.665531 11247 BufferMgr.cpp:375] ALLOCATION failed to find 300000256B free. Forcing Eviction. Eviction start 585938 Number pages requested 585938 Best Eviction Start Slab 2 CPU_MGR:0
I1017 20:18:50.835289 11250 BufferMgr.cpp:375] ALLOCATION failed to find 300000256B free. Forcing Eviction. Eviction start 2835210 Number pages requested 585938 Best Eviction Start Slab 2 CPU_MGR:0
I1017 20:18:55.884757 11246 BufferMgr.cpp:375] ALLOCATION failed to find 300000256B free. Forcing Eviction. Eviction start 3421148 Number pages requested 585938 Best Eviction Start Slab 2 CPU_MGR:0
I1017 20:19:07.918150 11248 BufferMgr.cpp:375] ALLOCATION failed to find 300000256B free. Forcing Eviction. Eviction start 4687503 Number pages requested 585938 Best Eviction Start Slab 0 CPU_MGR:0
I1017 20:19:12.988420 11247 BufferMgr.cpp:375] ALLOCATION failed to find 600000000B free. Forcing Eviction. Eviction start 5273441 Number pages requested 1171875 Best Eviction Start Slab 0 CPU_MGR:0
I1017 20:19:27.930819 11250 BufferMgr.cpp:375] ALLOCATION failed to find 150000128B free. Forcing Eviction. Eviction start 1757814 Number pages requested 292969 Best Eviction Start Slab 2 CPU_MGR:0
I1017 20:19:30.514434 11250 BufferMgr.cpp:375] ALLOCATION failed to find 300000256B free. Forcing Eviction. Eviction start 0 Number pages requested 585938 Best Eviction Start Slab 1 CPU_MGR:0
I1017 20:19:35.177902 11245 BufferMgr.cpp:375] ALLOCATION failed to find 300000256B free. Forcing Eviction. Eviction start 585938 Number pages requested 585938 Best Eviction Start Slab 1 CPU_MGR:0
I1017 20:19:39.440032 11245 BufferMgr.cpp:375] ALLOCATION failed to find 600000000B free. Forcing Eviction. Eviction start 1171876 Number pages requested 1171875 Best Eviction Start Slab 1 CPU_MGR:0
E1017 20:19:48.355754 11248 BufferMgr.cpp:370] ALLOCATION failed to find 600000000B throwing out of memory CPU_MGR:0
E1017 20:19:48.418704 11247 BufferMgr.cpp:370] ALLOCATION failed to find 600000000B throwing out of memory CPU_MGR:0
I1017 20:19:54.316274 11251 BufferMgr.cpp:375] ALLOCATION failed to find 251626496B free. Forcing Eviction. Eviction start 7617191 Number pages requested 491458 Best Eviction Start Slab 0 CPU_MGR:0
I1017 20:19:58.017437 11251 BufferMgr.cpp:375] ALLOCATION failed to find 503252992B free. Forcing Eviction. Eviction start 3080938 Number pages requested 982916 Best Eviction Start Slab 1 CPU_MGR:0
I1017 20:20:06.792656 11251 BufferMgr.cpp:375] ALLOCATION failed to find 503252992B free. Forcing Eviction. Eviction start 0 Number pages requested 982916 Best Eviction Start Slab 2 CPU_MGR:0
I1017 20:20:15.761013 11246 BufferMgr.cpp:375] ALLOCATION failed to find 300000256B free. Forcing Eviction. Eviction start 2050783 Number pages requested 585938 Best Eviction Start Slab 2 CPU_MGR:0
I1017 20:20:20.932322 11246 BufferMgr.cpp:375] ALLOCATION failed to find 150000128B free. Forcing Eviction. Eviction start 7031253 Number pages requested 292969 Best Eviction Start Slab 1 CPU_MGR:0
I1017 20:20:22.847684 11246 BufferMgr.cpp:375] ALLOCATION failed to find 300000256B free. Forcing Eviction. Eviction start 4300055 Number pages requested 585938 Best Eviction Start Slab 2 CPU_MGR:0
I1017 20:20:27.118314 11246 BufferMgr.cpp:375] ALLOCATION failed to find 600000000B free. Forcing Eviction. Eviction start 4687503 Number pages requested 1171875 Best Eviction Start Slab 0 CPU_MGR:0
E1017 20:20:36.944784 11246 BufferMgr.cpp:370] ALLOCATION failed to find 600000000B throwing out of memory CPU_MGR:0
I1017 20:20:36.953239 11250 BufferMgr.cpp:375] ALLOCATION failed to find 600000000B free. Forcing Eviction. Eviction start 4687503 Number pages requested 1171875 Best Eviction Start Slab 0 CPU_MGR:0
E1017 20:20:47.808964 11250 BufferMgr.cpp:370] ALLOCATION failed to find 600000000B throwing out of memory CPU_MGR:0
I1017 20:20:47.839965 11245 BufferMgr.cpp:375] ALLOCATION failed to find 600000000B free. Forcing Eviction. Eviction start 4687501 Number pages requested 1171875 Best Eviction Start Slab 1 CPU_MGR:0
I1017 20:20:59.088346 11249 BufferMgr.cpp:375] ALLOCATION failed to find 600000000B free. Forcing Eviction. Eviction start 2636721 Number pages requested 1171875 Best Eviction Start Slab 2 CPU_MGR:0
I1017 20:21:09.551349 11251 BufferMgr.cpp:375] ALLOCATION failed to find 251626496B free. Forcing Eviction. Eviction start 5859376 Number pages requested 491458 Best Eviction Start Slab 1 CPU_MGR:0
I1017 20:21:14.014973 11251 BufferMgr.cpp:375] ALLOCATION failed to find 503252992B free. Forcing Eviction. Eviction start 4300055 Number pages requested 982916 Best Eviction Start Slab 2 CPU_MGR:0
I1017 20:21:22.807905 11249 BufferMgr.cpp:375] ALLOCATION failed to find 600000000B free. Forcing Eviction. Eviction start 4687503 Number pages requested 1171875 Best Eviction Start Slab 0 CPU_MGR:0
E1017 20:21:33.186449 11249 BufferMgr.cpp:370] ALLOCATION failed to find 600000000B throwing out of memory CPU_MGR:0
I1017 20:21:33.188721 11251 BufferMgr.cpp:375] ALLOCATION failed to find 503252992B free. Forcing Eviction. Eviction start 5282971 Number pages requested 982916 Best Eviction Start Slab 2 CPU_MGR:0
I1017 20:21:42.195564 11251 BufferMgr.cpp:375] ALLOCATION failed to find 503252992B free. Forcing Eviction. Eviction start 0 Number pages requested 982916 Best Eviction Start Slab 0 CPU_MGR:0
I1017 20:21:50.880193 11251 BufferMgr.cpp:375] ALLOCATION failed to find 503252992B free. Forcing Eviction. Eviction start 3222658 Number pages requested 982916 Best Eviction Start Slab 0 CPU_MGR:0
I1017 20:22:04.150955 11245 BufferMgr.cpp:375] ALLOCATION failed to find 600000000B free. Forcing Eviction. Eviction start 982916 Number pages requested 1171875 Best Eviction Start Slab 2 CPU_MGR:0
I1017 20:22:13.288806 11251 BufferMgr.cpp:375] ALLOCATION failed to find 503252992B free. Forcing Eviction. Eviction start 2154791 Number pages requested 982916 Best Eviction Start Slab 2 CPU_MGR:0
I1017 20:22:21.842098 11245 BufferMgr.cpp:375] ALLOCATION failed to find 600000000B free. Forcing Eviction. Eviction start 4205574 Number pages requested 1171875 Best Eviction Start Slab 0 CPU_MGR:0
I1017 20:22:31.569769 11245 BufferMgr.cpp:375] ALLOCATION failed to find 600000000B free. Forcing Eviction. Eviction start 2343751 Number pages requested 1171875 Best Eviction Start Slab 1 CPU_MGR:0
I1017 20:22:40.674849 11245 BufferMgr.cpp:375] ALLOCATION failed to find 600000000B free. Forcing Eviction. Eviction start 3515626 Number pages requested 1171875 Best Eviction Start Slab 1 CPU_MGR:0
I1017 20:22:49.867633 11245 BufferMgr.cpp:375] ALLOCATION failed to find 600000000B free. Forcing Eviction. Eviction start 7031253 Number pages requested 1171875 Best Eviction Start Slab 1 CPU_MGR:0
E1017 20:22:59.365795  8275 MapDHandler.cpp:2306] Exception: Other error: code -16384
I1017 20:22:59.465265  8275 MapDHandler.cpp:321] User mapd disconnected from database mapd

```

**Comment by billmaimone** [comment]
Yes, that's the deal.  Currently the working set for the query needs to fit in GPU (or CPU) memory.  There's room for several straightforward optimizations for large projection queries that would improve the efficiency to allow much larger results.  Maybe this is a project for you one of these days, though I'd have in mind several other things ahead of this.

**Comment by fleapapa** [comment]
@billmaimone Would you mind giving a hint on one or two of the several straightforward optimizations in your mind? I feel it not simply an optimization, because current code acts more like a scatter-gather and lump-sum model while to enable it to stream larger datasets some parts of the code need to collaborate like synchronized pipelines or like coroutines.

**Comment by asuhan** [comment]
Support for this type of workload can be added without fundamental changes to the way we execute queries. We already have the ability to consume the input tables fragment-by-fragment; to minimize memory usage, we'd have to unpin the chunks in a fragment once it's been handled. Also, the output for each fragment can be written to disk right away instead of keeping it in memory. Ultimately it's a matter of priorities; I'm not sure of the relative importance of being the go-to system for what doesn't sound like an analytic workload vs other things we can improve.

Also, co-routines might be a convenient way to think about the problem, but ultimately they have to be lowered to native code and runtime support. I'm not sure we need that level of sophistication just to get reasonable overlap between data fetching and computation.

**Comment by fleapapa** [comment]
@asuhan. Agreed. Originally i thought the largest memory consumer is intermediate result, so i thought co-routine may help. Afterward i found the largest memory consumer is the file chunk buffers. So i've been probing usage of chunk buffers...

**Comment by asuhan** [comment]
You can't breakpoint into `RuntimeFunctions.bc` because that code is not directly linked into the server executable. Instead, the LLVM IR from it is inserted inline in the generated code. Also, the functions you've mentioned are actually replaced by real versions which are different between CPU and GPU when the query is compiled to native code.

**Comment by asuhan** [comment]
@fleapapa The intermediate memory for very big projection queries (of the select all rows and columns from the table kind) can be an issue as well. The good news is it can be optimized dramatically, to the point that a full co-routine model would be overkill on all but the most severely memory constrained systems. According to my back-of-the-envelope math, we could probably work with ~1 GB of memory without fundamental changes to the way the query executes, just by carefully implementing the ideas I've mentioned.

**Comment by fleapapa** [comment]
Quoted:
`We already have the ability to consume the input tables fragment-by-fragment; to minimize memory usage, we'd have to unpin the chunks in a fragment once it's been handled. `

As traced, the problem with this scheme is that for each fragment there're usually several threads (eg. 7 threads on my VM), each of which is responsible for loading a few columns. Since they run in parallel, an always-happening scenario with my test case is, as shown below, that there is always at least one thread that can't find a buffer big enough and can't wait for buffers unpinned by other threads before it throws OOM exception, which causes the query to fail. 

```
Slab St.Page   Pages  Touch
   0       0  585938 GAP:       0      0 PC:  1 USED - Chunk: 1,1,1,0,
   0  585938  585938 GAP:       0      1 PC:  1 USED - Chunk: 1,1,1,1,
   0 1171876  585938 GAP:       0      2 PC:  1 USED - Chunk: 1,1,2,1,
   0 1757814  585938 GAP:       0      3 PC:  1 USED - Chunk: 1,1,1,3,
   0 2343752  585938 GAP:       0      4 PC:  1 USED - Chunk: 1,1,2,3,
   0 2929690  585938 GAP:       0      5 PC:  1 USED - Chunk: 1,1,1,5,
   0 3515628  585938 GAP:       0      6 PC:  1 USED - Chunk: 1,1,2,5,
   0 4101566  585938 GAP:       0      7 PC:  1 USED - Chunk: 1,1,3,5,
   0 4687504  585938 GAP:       0      8 PC:  1 USED - Chunk: 1,1,1,2,
   0 5273442  585938 GAP:       0      9 PC:  1 USED - Chunk: 1,1,2,2,
   0 5859380  585938 GAP:       0     10 PC:  1 USED - Chunk: 1,1,3,2,
   0 6445318  585938 GAP:       0     11 PC:  1 USED - Chunk: 1,1,3,3,
   0 7031256  292969 GAP:       0     12 PC:  1 USED - Chunk: 1,1,4,3,
   0 7324225  585938 GAP:       0     13 PC:  1 USED - Chunk: 1,1,2,0,
   0 7910163  292969 GAP:       0     16 PC:  1 USED - Chunk: 1,1,4,1,
   0 8203132  185476 GAP:       0      0 FREE 
Slab St.Page   Pages  Touch
   1       0  585938 GAP:       0     14 PC:  1 USED - Chunk: 1,1,3,0,
   1  585938  585938 GAP:       0     15 PC:  1 USED - Chunk: 1,1,3,1,
   1 1171876  585938 GAP:       0     17 PC:  1 USED - Chunk: 1,1,5,1,
   1 1757814  491458 GAP:       0     18 PC:  1 USED - Chunk: 1,1,1,6,
   1 2249272  491458 GAP:       0     19 PC:  1 USED - Chunk: 1,1,2,6,
   1 2740730  491458 GAP:       0     20 PC:  1 USED - Chunk: 1,1,3,6,
   1 3232188  292969 GAP:       0     21 PC:  1 USED - Chunk: 1,1,4,0,
   1 3525157  585938 GAP:       0     22 PC:  1 USED - Chunk: 1,1,5,0,
   1 4111095 1171875 GAP:       0     23 PC:  1 USED - Chunk: 1,1,6,0,
   1 5282970 1171875 GAP:       0     24 PC:  1 USED - Chunk: 1,1,6,1,
   1 6454845  585938 GAP:       0     25 PC:  1 USED - Chunk: 1,1,5,3,
   1 7040783 1171875 GAP:       0     26 PC:  1 USED - Chunk: 1,1,6,3,
   1 8212658  175950 GAP:       0      0 FREE 
Slab St.Page   Pages  Touch
   2       0  245729 GAP:       0     27 PC:  1 USED - Chunk: 1,1,4,6,
   2  245729  491458 GAP:       0     28 PC:  1 USED - Chunk: 1,1,5,6,
   2  737187  292969 GAP:       0     29 PC:  1 USED - Chunk: 1,1,4,2,
   2 1030156 1171875 GAP:       0     30 PC:  1 USED - Chunk: 1,1,7,0,
   2 2202031  292969 GAP:       0     31 PC:  1 USED - Chunk: 1,1,8,0,
   2 2495000  585938 GAP:       0     32 PC:  1 USED - Chunk: 1,1,9,0,
   2 3080938 1171875 GAP:       0     33 PC:  1 USED - Chunk: 1,1,10,0,
   2 4252813  585938 GAP:       0     34 PC:  1 USED - Chunk: 1,1,1,4,
   2 4838751  982916 GAP:       0     35 PC:  1 USED - Chunk: 1,1,6,6,
   2 5821667  585938 GAP:       0     36 PC:  1 USED - Chunk: 1,1,5,2,
   2 6407605  225796 GAP:       0      0 FREE 

```

**Comment by asuhan** [comment]
Yes, we'd have to throttle the execution (a memory quota per thread), which ultimately means that our optimization on the query execution side is less useful for this type of workloads. This is the reason we didn't really focus on them -- we have less of a comparative advantage for low-memory scenarios vs the ones we've already focused on.

**Comment by fleapapa** [comment]
I guess for this kind of work load users usually look for _entirety_ of query result than query performance.

For example, when i backup/copy my BIG tables, i am more concerned about whether all my data is written to destination storage than about how long it will take.

**Comment by asuhan** [comment]
Yes, backup is a potential application for it. At the same time, backup can (and probably should) be implemented as a separate feature, it doesn't need to be in the roundabout way of running a `SELECT *` query. At the same time, there's definitely room to better accommodate certain non-analytic queries for well-defined purposes.

**Comment by fleapapa** [comment]
I might have missing another application -- ETL :)

ETL jobs usually query/aggregate multiple BIG tables to generate a BIG datasets for offline analytics and model training.

**Comment by fleapapa** [comment]
@asuhan,

`Yes, we'd have to throttle the execution (a memory quota per thread), `

Per your hint, i took an easier path. Instead of throttling memory usage per thread, i sequentially ran the threads. This time, all fragments were fetched, but then execution halted at GroupByAndAggregate.cpp:138. To proceed it needs a big "group by" buffer of 60 GB (the table uses 47 GB disk space). I could go get 4 x 16GB DIMM but what if my table is 1TB or larger? I don't think my server can install TBs or even 1TB memory...

**Comment by asuhan** [comment]
That's where the dramatical reduction of memory usage for projection output comes into play. For big projection queries, a bitmap layout would be orders of magnitude smaller, you wouldn't need anything close to 60GB. More like 600MB.

**Comment by fleapapa** [comment]
Is the "dramatical reduction" in the code? A quick pointer or hint will be appreciated.

**Comment by asuhan** [comment]
No, it's not yet, but we know how to build it and I think we also have an internal task for it. It's not particularly difficult, but it touches all the executor parts (some code generation, the result set etc).

**Comment by fleapapa** [comment]
@asuhan. I have a long question of Executor::executePlanWithoutGroupBy().

My query is "COPY (SELECT * FROM trips) TO 'trips.csv'". The table trips has ~0.5B rows. From gdb, after the first fragment (75000000 rows x 14 columns) was read into memory, the first call of executePlanWithoutGroupBy() came out with a result set which has only one value:

```
(gdb) p reduced_outs
$28 = std::vector of length 1, capacity 1 = {75000000}
```

I thought the result set would be a 75M x 14 block of values. My query has no aggregation, so i'm confused why the function ends up with only a single value. Please shed some light here.

**Comment by asuhan** [comment]
I doubt that method should be even hit by the query. Are you certain that the source is not newer / older than the executable? Also, what's in the output `trips.csv` file, does the operation work?

**Comment by fleapapa** [comment]
No, the operation doesn't work yet. The file trips.csv isn't even created, because after 7th call of the method (because the table has 7 fragments), the code always ends up with a failure to allocate 60GB memory.

The method was indeed hit, as shown below.

```
(gdb) bt
#0  Executor::executePlanWithoutGroupBy (this=0x7fffd401edb0, ra_exe_unit=..., compilation_result=..., hoist_literals=true, results=..., target_exprs=std::vector of length 1, capacity 1 = {...}, 
    device_type=ExecutorDeviceType::CPU, col_buffers=std::vector of length 1, capacity 1 = {...}, query_exe_context=0x7fffd40dc620, num_rows=std::vector of length 1, capacity 1 = {...}, 
    frag_offsets=std::vector of length 1, capacity 1 = {...}, frag_stride=1, data_mgr=0x388f6e0, device_id=0, start_rowid=0, num_tables=1, render_allocator_map=0x0) at /data/github/mapd-core/QueryEngine/Execute.cpp:1845
#1  0x00000000012cc3d3 in Executor::ExecutionDispatch::runImpl (this=0x7fffdf101f90, chosen_device_type=ExecutorDeviceType::CPU, chosen_device_id=0, options=..., frag_ids=std::vector of length 1, capacity 1 = {...}, ctx_idx=0, 
    rowid_lookup_key=-1) at /data/github/mapd-core/QueryEngine/ExecutionDispatch.cpp:333
#2  0x00000000012cdcf0 in Executor::ExecutionDispatch::run (this=0x7fffdf101f90, chosen_device_type=ExecutorDeviceType::CPU, chosen_device_id=0, options=..., frag_ids=std::vector of length 1, capacity 1 = {...}, ctx_idx=0, 
    rowid_lookup_key=-1) at /data/github/mapd-core/QueryEngine/ExecutionDispatch.cpp:513
#3  0x000000000123b8ba in Executor::<lambda(ExecutorDeviceType, int, const std::vector<std::pair<int, std::vector<long unsigned int, std::allocator<long unsigned int> > >, std::allocator<std::pair<int, std::vector<long unsigned int, std::allocator<long unsigned int> > > > >&, size_t, int64_t)>::operator()(ExecutorDeviceType, int, const std::vector<std::pair<int, std::vector<unsigned long, std::allocator<unsigned long> > >, std::allocator<std::pair<int, std::vector<unsigned long, std::allocator<unsigned long> > > > > &, size_t, int64_t) const (__closure=0x7fffd4221220, chosen_device_type=ExecutorDeviceType::CPU, chosen_device_id=0, 
    frag_ids=std::vector of length 1, capacity 1 = {...}, ctx_idx=0, rowid_lookup_key=-1) at /data/github/mapd-core/QueryEngine/Execute.cpp:868
#4  0x0000000001273b9f in std::_Function_handler<void(ExecutorDeviceType, int, const std::vector<std::pair<int, std::vector<long unsigned int, std::allocator<long unsigned int> > >, std::allocator<std::pair<int, std::vector<long unsigned int, std::allocator<long unsigned int> > > > >&, long unsigned int, long int), Executor::executeWorkUnit(int32_t*, size_t&, bool, const std::vector<InputTableInfo>&, const RelAlgExecutionUnit&, const CompilationOptions&, const ExecutionOptions&, const Catalog_Namespace::Catalog&, std::shared_ptr<RowSetMemoryOwner>, RenderAllocatorMap*, bool)::<lambda(ExecutorDeviceType, int, const std::vector<std::pair<int, std::vector<long unsigned int, std::allocator<long unsigned int> > >, std::allocator<std::pair<int, std::vector<long unsigned int, std::allocator<long unsigned int> > > > >&, size_t, int64_t)> >::_M_invoke(const std::_Any_data &, <unknown type in /mnt/mapd-core/bin/mapd_server, CU 0xa37f35, DIE 0xbc6b51>, <unknown type in /mnt/mapd-core/bin/mapd_server, CU 0xa37f35, DIE 0xb7f2bf>, const std::vector<std::pair<int, std::vector<unsigned long, std::allocator<unsigned long> > >, std::allocator<std::pair<int, std::vector<unsigned long, std::allocator<unsigned long> > > > > &, <unknown type in /mnt/mapd-core/bin/mapd_server, CU 0xa37f35, DIE 0xb56ed8>, <unknown type in /mnt/mapd-core/bin/mapd_server, CU 0xa37f35, DIE 0xb5a203>) (__functor=..., __args#0=<unknown type in /mnt/mapd-core/bin/mapd_server, CU 0xa37f35, DIE 0xbc6b51>, __args#1=<unknown type in /mnt/mapd-core/bin/mapd_server, CU 0xa37f35, DIE 0xb7f2bf>, 
    __args#2=std::vector of length 1, capacity 1 = {...}, __args#3=<unknown type in /mnt/mapd-core/bin/mapd_server, CU 0xa37f35, DIE 0xb56ed8>, 
    __args#4=<unknown type in /mnt/mapd-core/bin/mapd_server, CU 0xa37f35, DIE 0xb5a203>) at /usr/include/c++/6/functional:1731
#5  0x0000000001283716 in std::function<void (ExecutorDeviceType, int, std::vector<std::pair<int, std::vector<unsigned long, std::allocator<unsigned long> > >, std::allocator<std::pair<int, std::vector<unsigned long, std::allocator<unsigned long> > > > > const&, unsigned long, long)>::operator()(ExecutorDeviceType, int, std::vector<std::pair<int, std::vector<unsigned long, std::allocator<unsigned long> > >, std::allocator<std::pair<int, std::vector<unsigned long, std::allocator<unsigned long> > > > > const&, unsigned long, long) const (this=0x7fffdf101e20, __args#0=ExecutorDeviceType::CPU, __args#1=0, __args#2=std::vector of length 1, capacity 1 = {...}, __args#3=0, 
    __args#4=-1) at /usr/include/c++/6/functional:2127
#6  0x0000000001240552 in Executor::dispatchFragments(std::function<void (ExecutorDeviceType, int, std::vector<std::pair<int, std::vector<unsigned long, std::allocator<unsigned long> > >, std::allocator<std::pair<int, std::vector<unsigned long, std::allocator<unsigned long> > > > > const&, unsigned long, long)>, Executor::ExecutionDispatch const&, ExecutionOptions const&, bool, std::map<int, std::deque<Fragmenter_Namespace::FragmentInfo, std::allocator<Fragmenter_Namespace::FragmentInfo> > const*, std::less<int>, std::allocator<std::pair<int const, std::deque<Fragmenter_Namespace::FragmentInfo, std::allocator<Fragmenter_Namespace::FragmentInfo> > const*> > >&, unsigned long, std::condition_variable&, std::mutex&, std::unordered_set<int, std::hash<int>, std::equal_to<int>, std::allocator<int> >&, int&) (this=0x7fffd401edb0, dispatch=..., execution_dispatch=..., eo=..., is_agg=true, 
    selected_tables_fragments=std::map with 1 element = {...}, context_count=14, scheduler_cv=..., scheduler_mutex=..., available_gpus=std::unordered_set with 0 elements, available_cpus=@0x7fffdf101ba4: 14)
    at /data/github/mapd-core/QueryEngine/Execute.cpp:1296
#7  0x000000000123c60e in Executor::executeWorkUnit (this=0x7fffd401edb0, error_code=0x7fffdf1024dc, max_groups_buffer_entry_guess=@0x7fffdf1024e0: 1, is_agg=true, query_infos=std::vector of length 1, capacity 1 = {...}, 
    ra_exe_unit=..., co=..., options=..., cat=..., row_set_mem_owner=std::shared_ptr (count 5, weak 0) 0x7fffd4033fe0, render_allocator_map=0x0, has_cardinality_estimation=false)
    at /data/github/mapd-core/QueryEngine/Execute.cpp:907
#8  0x0000000001380399 in RelAlgExecutor::getFilteredCountAll (this=0x7fffdf103720, work_unit=..., is_agg=true, co=..., eo=...) at /data/github/mapd-core/QueryEngine/RelAlgExecutor.cpp:1417
#9  0x000000000137f57b in RelAlgExecutor::executeWorkUnit (this=0x7fffdf103720, work_unit=..., targets_meta=std::vector of length 14, capacity 14 = {...}, is_agg=false, co=..., eo=..., render_info=0x0, queue_time_ms=0)
    at /data/github/mapd-core/QueryEngine/RelAlgExecutor.cpp:1309
#10 0x000000000137d522 in RelAlgExecutor::executeProject (this=0x7fffdf103720, project=0x7fffd4032220, co=..., eo=..., render_info=0x0, queue_time_ms=0) at /data/github/mapd-core/QueryEngine/RelAlgExecutor.cpp:1036
#11 0x0000000001377d96 in RelAlgExecutor::executeRelAlgStep (this=0x7fffdf103720, i=0, exec_descs=std::vector of length 1, capacity 1 = {...}, co=..., eo=..., render_info=0x0, queue_time_ms=0)
    at /data/github/mapd-core/QueryEngine/RelAlgExecutor.cpp:379
#12 0x0000000001376346 in RelAlgExecutor::executeRelAlgSeq (this=0x7fffdf103720, exec_descs=std::vector of length 1, capacity 1 = {...}, co=..., eo=..., render_info=0x0, queue_time_ms=0)
    at /data/github/mapd-core/QueryEngine/RelAlgExecutor.cpp:231
#13 0x0000000001374bb3 in RelAlgExecutor::executeRelAlgQueryNoRetry (this=0x7fffdf103720, 
    query_ra="{\n  \"rels\": [\n    {\n      \"id\": \"0\",\n      \"relOp\": \"EnumerableTableScan\",\n      \"fieldNames\": [\n        \"medallion\",\n        \"hack_license\",\n        \"vendor_id\",\n        \"rate_code_id\",\n        \"stor"..., co=..., eo=..., render_info=0x0) at /data/github/mapd-core/QueryEngine/RelAlgExecutor.cpp:86
#14 0x000000000137451e in RelAlgExecutor::executeRelAlgQuery (this=0x7fffdf103720, 
    query_ra="{\n  \"rels\": [\n    {\n      \"id\": \"0\",\n      \"relOp\": \"EnumerableTableScan\",\n      \"fieldNames\": [\n        \"medallion\",\n        \"hack_license\",\n        \"vendor_id\",\n        \"rate_code_id\",\n        \"stor"..., co=..., eo=..., render_info=0x0) at /data/github/mapd-core/QueryEngine/RelAlgExecutor.cpp:40
#15 0x0000000001183564 in Parser::getResultRows (session=..., select_stmt="select * from trips;", targets=std::vector of length 0, capacity 0) at /data/github/mapd-core/Parser/ParserNode.cpp:1927
#16 0x000000000118c703 in Parser::ExportQueryStmt::execute (this=0x7fffd4016340, session=...) at /data/github/mapd-core/Parser/ParserNode.cpp:2760
---Type <return> to continue, or q <return> to quit---q
Quit

```

**Comment by fleapapa** [comment]
The executable should be in sync with the source code, because each time i rebuilt the executable gdb notified me of symbol reload.

**Comment by asuhan** [comment]
Ok, now I understand what's going on. As you can see in the stack trace, that first `ResultSet` comes from a `RelAlgExecutor::getFilteredCountAll` call (and it looks correct). For projection queries on big tables, we first run a `COUNT(*)` query (with all the filters applied, if any) to find out how big the result should be. So there's nothing concerning / incorrect here.

**Comment by fleapapa** [comment]
Why doesn't it simply run SELECT * directly?  Otherwise (especially with tight memory) when it does it needs to reload the fragment from disk.

**Comment by fleapapa** [comment]
Or, is it complicated (for me) to bypass getFilteredCountAll?

**Comment by asuhan** [comment]
1. No, we don't need to load data from disk except for the columns which are used for the filter (none in this case), for the rest we only need the meta data.
2. Having to start with a uneducated guess and keep reallocating is very expensive compared to running a cheap `COUNT(*)` query.
3. No, it's not complicated at all, but for the reasons above you don't want to.

**Comment by fleapapa** [comment]
I don't understand 1. My VM has only 14GB free memory. The table needs at least 48GB space, so each fragment needs at least 6GB. That means after 3rd call of the method, the first fragment would have been evicted.

**Comment by asuhan** [comment]
On what grounds do you think that a `SELECT COUNT(*) FROM trips` needs to load anything but some metadata in the memory? Is there evidence for it?

**Comment by fleapapa** [comment]
From the pages loaded and timestamps, you can see the server worked hard to read 48GB data into memory (by many evictions). That's why i think it needs to reload the fragments if the code does not directly perform my query on the first chance.

```
I1020 21:25:16.984474 23391 MapDHandler.cpp:2206] copy (select * from trips) to '/mnt/tmp/trips.csv';
I1020 21:25:16.987905 23391 MapDHandler.cpp:2234] passing query to legacy processor
I1020 21:25:16.996446 23391 Calcite.cpp:247] User mapd catalog mapd sql 'select * from trips;'
I1020 21:28:44.089020 23394 FileMgr.cpp:173] Completed Reading table's file metadata, Elasped time : 206148ms Epoch: 125234 files read: 506 table location: '/mnt/trip/mapd/data/mapd_data/table_1_1/'
I1020 21:28:56.028679 23394 Catalog.cpp:1511] Instantiating Fragmenter for table trips took 218122ms
I1020 21:28:56.243412 23394 Catalog.cpp:1557] Time to load Dictionary 1 was 214ms
I1020 21:28:56.386852 23394 Catalog.cpp:1557] Time to load Dictionary 2 was 143ms
I1020 21:28:56.520334 23394 Catalog.cpp:1557] Time to load Dictionary 3 was 133ms
I1020 21:28:56.630808 23394 Catalog.cpp:1557] Time to load Dictionary 4 was 110ms
I1020 21:28:56.929828 23391 Calcite.cpp:260] Time in Thrift 28 (ms), Time in Java Calcite server 219905 (ms)
I1020 21:28:57.484974 23391 BufferMgr.cpp:270] ALLOCATION slab of 8388608 pages (4294967296B) created in 0 ms CPU_MGR:0
I1020 21:29:41.467334 23391 BufferMgr.cpp:270] ALLOCATION slab of 8388608 pages (4294967296B) created in 12 ms CPU_MGR:0
I1020 21:30:02.858394 23391 ExecutionDispatch.cpp:375] 1 results so far
I1020 21:30:02.858515 23391 ExecutionDispatch.cpp:380] 	 frag 0: 1 rows, 1 cols
I1020 21:30:28.046356 23391 BufferMgr.cpp:270] ALLOCATION slab of 6633420 pages (3396311040B) created in 0 ms CPU_MGR:0
I1020 21:31:20.059157 23391 BufferMgr.cpp:379] ALLOCATION failed to find 600000000B free. Forcing Eviction. Eviction start 0 Number pages requested 1171875 Best Eviction Start Slab 0 CPU_MGR:0
I1020 21:31:29.099066 23391 ExecutionDispatch.cpp:375] 2 results so far
I1020 21:31:29.099279 23391 ExecutionDispatch.cpp:380] 	 frag 0: 1 rows, 1 cols
I1020 21:31:29.099329 23391 ExecutionDispatch.cpp:380] 	 frag 1: 1 rows, 1 cols
I1020 21:31:34.358129 23391 BufferMgr.cpp:379] ALLOCATION failed to find 300000256B free. Forcing Eviction. Eviction start 1171875 Number pages requested 585938 Best Eviction Start Slab 0 CPU_MGR:0
I1020 21:31:39.039259 23391 BufferMgr.cpp:379] ALLOCATION failed to find 300000256B free. Forcing Eviction. Eviction start 1757813 Number pages requested 585938 Best Eviction Start Slab 0 CPU_MGR:0
I1020 21:31:43.808163 23391 BufferMgr.cpp:379] ALLOCATION failed to find 300000256B free. Forcing Eviction. Eviction start 2636720 Number pages requested 585938 Best Eviction Start Slab 0 CPU_MGR:0
I1020 21:31:47.103474 23391 BufferMgr.cpp:379] ALLOCATION failed to find 600000000B free. Forcing Eviction. Eviction start 3222658 Number pages requested 1171875 Best Eviction Start Slab 0 CPU_MGR:0
I1020 21:31:56.545462 23391 BufferMgr.cpp:379] ALLOCATION failed to find 600000000B free. Forcing Eviction. Eviction start 4394533 Number pages requested 1171875 Best Eviction Start Slab 0 CPU_MGR:0
I1020 21:32:07.710490 23391 BufferMgr.cpp:379] ALLOCATION failed to find 300000256B free. Forcing Eviction. Eviction start 5859377 Number pages requested 585938 Best Eviction Start Slab 0 CPU_MGR:0
I1020 21:32:11.881705 23391 BufferMgr.cpp:379] ALLOCATION failed to find 600000000B free. Forcing Eviction. Eviction start 6445315 Number pages requested 1171875 Best Eviction Start Slab 0 CPU_MGR:0
I1020 21:32:21.657599 23391 BufferMgr.cpp:379] ALLOCATION failed to find 600000000B free. Forcing Eviction. Eviction start 0 Number pages requested 1171875 Best Eviction Start Slab 1 CPU_MGR:0
I1020 21:32:31.904186 23391 BufferMgr.cpp:379] ALLOCATION failed to find 600000000B free. Forcing Eviction. Eviction start 1171875 Number pages requested 1171875 Best Eviction Start Slab 1 CPU_MGR:0
I1020 21:32:41.506430 23391 BufferMgr.cpp:379] ALLOCATION failed to find 600000000B free. Forcing Eviction. Eviction start 2343750 Number pages requested 1171875 Best Eviction Start Slab 1 CPU_MGR:0
I1020 21:32:51.118618 23391 BufferMgr.cpp:379] ALLOCATION failed to find 600000000B free. Forcing Eviction. Eviction start 3515625 Number pages requested 1171875 Best Eviction Start Slab 1 CPU_MGR:0
I1020 21:33:01.206462 23391 ExecutionDispatch.cpp:375] 3 results so far
I1020 21:33:01.208508 23391 ExecutionDispatch.cpp:380] 	 frag 0: 1 rows, 1 cols
I1020 21:33:01.208561 23391 ExecutionDispatch.cpp:380] 	 frag 1: 1 rows, 1 cols
I1020 21:33:01.208590 23391 ExecutionDispatch.cpp:380] 	 frag 2: 1 rows, 1 cols
I1020 21:33:06.430074 23391 BufferMgr.cpp:379] ALLOCATION failed to find 300000256B free. Forcing Eviction. Eviction start 4687500 Number pages requested 585938 Best Eviction Start Slab 1 CPU_MGR:0
I1020 21:33:10.293550 23391 BufferMgr.cpp:379] ALLOCATION failed to find 300000256B free. Forcing Eviction. Eviction start 5273438 Number pages requested 585938 Best Eviction Start Slab 1 CPU_MGR:0
I1020 21:33:15.068744 23391 BufferMgr.cpp:379] ALLOCATION failed to find 300000256B free. Forcing Eviction. Eviction start 6152345 Number pages requested 585938 Best Eviction Start Slab 1 CPU_MGR:0
I1020 21:33:18.478672 23391 BufferMgr.cpp:379] ALLOCATION failed to find 600000000B free. Forcing Eviction. Eviction start 0 Number pages requested 1171875 Best Eviction Start Slab 2 CPU_MGR:0
I1020 21:33:28.146893 23391 BufferMgr.cpp:379] ALLOCATION failed to find 600000000B free. Forcing Eviction. Eviction start 6738283 Number pages requested 1171875 Best Eviction Start Slab 1 CPU_MGR:0
I1020 21:33:40.015305 23391 BufferMgr.cpp:379] ALLOCATION failed to find 300000256B free. Forcing Eviction. Eviction start 1171875 Number pages requested 585938 Best Eviction Start Slab 2 CPU_MGR:0
I1020 21:33:44.316450 23391 BufferMgr.cpp:379] ALLOCATION failed to find 600000000B free. Forcing Eviction. Eviction start 1757813 Number pages requested 1171875 Best Eviction Start Slab 2 CPU_MGR:0
I1020 21:33:53.960752 23391 BufferMgr.cpp:379] ALLOCATION failed to find 600000000B free. Forcing Eviction. Eviction start 2929688 Number pages requested 1171875 Best Eviction Start Slab 2 CPU_MGR:0
I1020 21:34:03.610249 23391 BufferMgr.cpp:379] ALLOCATION failed to find 600000000B free. Forcing Eviction. Eviction start 4101563 Number pages requested 1171875 Best Eviction Start Slab 2 CPU_MGR:0
I1020 21:34:13.247299 23391 BufferMgr.cpp:379] ALLOCATION failed to find 600000000B free. Forcing Eviction. Eviction start 0 Number pages requested 1171875 Best Eviction Start Slab 0 CPU_MGR:0
I1020 21:34:23.719784 23391 BufferMgr.cpp:379] ALLOCATION failed to find 600000000B free. Forcing Eviction. Eviction start 5273438 Number pages requested 1171875 Best Eviction Start Slab 2 CPU_MGR:0
I1020 21:34:34.328130 23391 ExecutionDispatch.cpp:375] 4 results so far
I1020 21:34:34.330106 23391 ExecutionDispatch.cpp:380] 	 frag 0: 1 rows, 1 cols
I1020 21:34:34.330166 23391 ExecutionDispatch.cpp:380] 	 frag 1: 1 rows, 1 cols
I1020 21:34:34.330194 23391 ExecutionDispatch.cpp:380] 	 frag 2: 1 rows, 1 cols
I1020 21:34:34.330219 23391 ExecutionDispatch.cpp:380] 	 frag 3: 1 rows, 1 cols
I1020 21:34:34.330456 23391 BufferMgr.cpp:379] ALLOCATION failed to find 300000256B free. Forcing Eviction. Eviction start 1171875 Number pages requested 585938 Best Eviction Start Slab 0 CPU_MGR:0
I1020 21:34:39.465829 23391 BufferMgr.cpp:379] ALLOCATION failed to find 300000256B free. Forcing Eviction. Eviction start 1757813 Number pages requested 585938 Best Eviction Start Slab 0 CPU_MGR:0
I1020 21:34:43.156139 23391 BufferMgr.cpp:379] ALLOCATION failed to find 300000256B free. Forcing Eviction. Eviction start 2343751 Number pages requested 585938 Best Eviction Start Slab 0 CPU_MGR:0
I1020 21:34:48.416127 23391 BufferMgr.cpp:379] ALLOCATION failed to find 300000256B free. Forcing Eviction. Eviction start 3222658 Number pages requested 585938 Best Eviction Start Slab 0 CPU_MGR:0
I1020 21:34:51.833508 23391 BufferMgr.cpp:379] ALLOCATION failed to find 600000000B free. Forcing Eviction. Eviction start 3808596 Number pages requested 1171875 Best Eviction Start Slab 0 CPU_MGR:0
I1020 21:35:01.579792 23391 BufferMgr.cpp:379] ALLOCATION failed to find 600000000B free. Forcing Eviction. Eviction start 4980471 Number pages requested 1171875 Best Eviction Start Slab 0 CPU_MGR:0
I1020 21:35:14.060984 23391 BufferMgr.cpp:379] ALLOCATION failed to find 300000256B free. Forcing Eviction. Eviction start 6445315 Number pages requested 585938 Best Eviction Start Slab 0 CPU_MGR:0
I1020 21:35:18.560178 23391 BufferMgr.cpp:379] ALLOCATION failed to find 600000000B free. Forcing Eviction. Eviction start 0 Number pages requested 1171875 Best Eviction Start Slab 1 CPU_MGR:0
I1020 21:35:28.192518 23391 BufferMgr.cpp:379] ALLOCATION failed to find 600000000B free. Forcing Eviction. Eviction start 1171875 Number pages requested 1171875 Best Eviction Start Slab 1 CPU_MGR:0
I1020 21:35:38.072480 23391 BufferMgr.cpp:379] ALLOCATION failed to find 600000000B free. Forcing Eviction. Eviction start 2343750 Number pages requested 1171875 Best Eviction Start Slab 1 CPU_MGR:0
I1020 21:35:48.236078 23391 BufferMgr.cpp:379] ALLOCATION failed to find 600000000B free. Forcing Eviction. Eviction start 3515625 Number pages requested 1171875 Best Eviction Start Slab 1 CPU_MGR:0
I1020 21:35:58.605482 23391 BufferMgr.cpp:379] ALLOCATION failed to find 600000000B free. Forcing Eviction. Eviction start 7031253 Number pages requested 1171875 Best Eviction Start Slab 0 CPU_MGR:0
I1020 21:36:09.166383 23391 ExecutionDispatch.cpp:375] 5 results so far
I1020 21:36:09.168303 23391 ExecutionDispatch.cpp:380] 	 frag 0: 1 rows, 1 cols
I1020 21:36:09.168359 23391 ExecutionDispatch.cpp:380] 	 frag 1: 1 rows, 1 cols
I1020 21:36:09.168386 23391 ExecutionDispatch.cpp:380] 	 frag 2: 1 rows, 1 cols
I1020 21:36:09.168411 23391 ExecutionDispatch.cpp:380] 	 frag 3: 1 rows, 1 cols
I1020 21:36:09.168436 23391 ExecutionDispatch.cpp:380] 	 frag 4: 1 rows, 1 cols
I1020 21:36:09.168648 23391 BufferMgr.cpp:379] ALLOCATION failed to find 300000256B free. Forcing Eviction. Eviction start 4687500 Number pages requested 585938 Best Eviction Start Slab 1 CPU_MGR:0
I1020 21:36:14.405619 23391 BufferMgr.cpp:379] ALLOCATION failed to find 300000256B free. Forcing Eviction. Eviction start 5273438 Number pages requested 585938 Best Eviction Start Slab 1 CPU_MGR:0
I1020 21:36:17.998975 23391 BufferMgr.cpp:379] ALLOCATION failed to find 300000256B free. Forcing Eviction. Eviction start 5859376 Number pages requested 585938 Best Eviction Start Slab 1 CPU_MGR:0
I1020 21:36:22.773504 23391 BufferMgr.cpp:379] ALLOCATION failed to find 300000256B free. Forcing Eviction. Eviction start 0 Number pages requested 585938 Best Eviction Start Slab 2 CPU_MGR:0
I1020 21:36:25.936898 23391 BufferMgr.cpp:379] ALLOCATION failed to find 600000000B free. Forcing Eviction. Eviction start 6738283 Number pages requested 1171875 Best Eviction Start Slab 1 CPU_MGR:0
I1020 21:36:35.969362 23391 BufferMgr.cpp:379] ALLOCATION failed to find 600000000B free. Forcing Eviction. Eviction start 585938 Number pages requested 1171875 Best Eviction Start Slab 2 CPU_MGR:0
I1020 21:36:46.441886 23391 BufferMgr.cpp:379] ALLOCATION failed to find 150000128B free. Forcing Eviction. Eviction start 7910158 Number pages requested 292969 Best Eviction Start Slab 1 CPU_MGR:0
I1020 21:36:48.611456 23391 BufferMgr.cpp:379] ALLOCATION failed to find 300000256B free. Forcing Eviction. Eviction start 1757813 Number pages requested 585938 Best Eviction Start Slab 2 CPU_MGR:0
I1020 21:36:53.013679 23391 BufferMgr.cpp:379] ALLOCATION failed to find 600000000B free. Forcing Eviction. Eviction start 2343751 Number pages requested 1171875 Best Eviction Start Slab 2 CPU_MGR:0
I1020 21:37:02.910850 23391 BufferMgr.cpp:379] ALLOCATION failed to find 600000000B free. Forcing Eviction. Eviction start 3515626 Number pages requested 1171875 Best Eviction Start Slab 2 CPU_MGR:0
I1020 21:37:13.192278 23391 BufferMgr.cpp:379] ALLOCATION failed to find 600000000B free. Forcing Eviction. Eviction start 0 Number pages requested 1171875 Best Eviction Start Slab 0 CPU_MGR:0
I1020 21:37:23.426311 23391 BufferMgr.cpp:379] ALLOCATION failed to find 600000000B free. Forcing Eviction. Eviction start 4687501 Number pages requested 1171875 Best Eviction Start Slab 2 CPU_MGR:0
I1020 21:37:33.419581 23391 BufferMgr.cpp:379] ALLOCATION failed to find 600000000B free. Forcing Eviction. Eviction start 1171875 Number pages requested 1171875 Best Eviction Start Slab 0 CPU_MGR:0
I1020 21:37:43.788893 23391 ExecutionDispatch.cpp:375] 6 results so far
I1020 21:37:43.791416 23391 ExecutionDispatch.cpp:380] 	 frag 0: 1 rows, 1 cols
I1020 21:37:43.791553 23391 ExecutionDispatch.cpp:380] 	 frag 1: 1 rows, 1 cols
I1020 21:37:43.791599 23391 ExecutionDispatch.cpp:380] 	 frag 2: 1 rows, 1 cols
I1020 21:37:43.791626 23391 ExecutionDispatch.cpp:380] 	 frag 3: 1 rows, 1 cols
I1020 21:37:43.791651 23391 ExecutionDispatch.cpp:380] 	 frag 4: 1 rows, 1 cols
I1020 21:37:43.791674 23391 ExecutionDispatch.cpp:380] 	 frag 5: 1 rows, 1 cols
I1020 21:37:48.209626 23391 BufferMgr.cpp:379] ALLOCATION failed to find 251626496B free. Forcing Eviction. Eviction start 2343750 Number pages requested 491458 Best Eviction Start Slab 0 CPU_MGR:0
I1020 21:37:51.032223 23391 BufferMgr.cpp:379] ALLOCATION failed to find 251626496B free. Forcing Eviction. Eviction start 2835208 Number pages requested 491458 Best Eviction Start Slab 0 CPU_MGR:0
I1020 21:37:55.281162 23391 BufferMgr.cpp:379] ALLOCATION failed to find 251626496B free. Forcing Eviction. Eviction start 3572395 Number pages requested 491458 Best Eviction Start Slab 0 CPU_MGR:0
I1020 21:37:58.007580 23391 BufferMgr.cpp:379] ALLOCATION failed to find 503252992B free. Forcing Eviction. Eviction start 4063853 Number pages requested 982916 Best Eviction Start Slab 0 CPU_MGR:0
I1020 21:38:16.318179 23391 BufferMgr.cpp:379] ALLOCATION failed to find 251626496B free. Forcing Eviction. Eviction start 6029685 Number pages requested 491458 Best Eviction Start Slab 0 CPU_MGR:0
I1020 21:38:20.007076 23391 BufferMgr.cpp:379] ALLOCATION failed to find 503252992B free. Forcing Eviction. Eviction start 0 Number pages requested 982916 Best Eviction Start Slab 1 CPU_MGR:0
I1020 21:38:28.331648 23391 BufferMgr.cpp:379] ALLOCATION failed to find 503252992B free. Forcing Eviction. Eviction start 982916 Number pages requested 982916 Best Eviction Start Slab 1 CPU_MGR:0
I1020 21:38:37.336382 23391 BufferMgr.cpp:379] ALLOCATION failed to find 503252992B free. Forcing Eviction. Eviction start 1965832 Number pages requested 982916 Best Eviction Start Slab 1 CPU_MGR:0
I1020 21:38:46.038153 23391 BufferMgr.cpp:379] ALLOCATION failed to find 503252992B free. Forcing Eviction. Eviction start 2948748 Number pages requested 982916 Best Eviction Start Slab 1 CPU_MGR:0
I1020 21:38:54.665050 23391 BufferMgr.cpp:379] ALLOCATION failed to find 503252992B free. Forcing Eviction. Eviction start 6521143 Number pages requested 982916 Best Eviction Start Slab 0 CPU_MGR:0

```

**Comment by asuhan** [comment]
Did the 48GB of reading happened *before* `RelAlgExecutor::getFilteredCountAll`? Because we definitely do a lot of reading for the actual `SELECT *` query.

**Comment by asuhan** [comment]
To sum up:

1. The amount of memory required by `RelAlgExecutor::getFilteredCountAll` is negligible -- in your case, it only needs to load the metadata for the fragments.
2. It doesn't hurt to know how many rows will the original query yield. The fact that we think we need 60GB for a mere 75M rows is what needs to be fixed. The `RelAlgExecutor::getFilteredCountAll` correctly and cheaply tells us that we need space for 75M rows -- how much space we allocate for each row in the case of a projection query is the root problem.

**Comment by fleapapa** [comment]
For each call of the, it does, because in Executor::ExecutionDispatch::runImpl the call to fetchChunks the fragment is before the call to executePlanWithoutGroupBy. 

One thing i'm sure is that it took quite some time before each entry to executePlanWithoutGroupBy but comes out with only a single integer value (=75k).

**Comment by asuhan** [comment]
I don't think continuing this conversation is productive. Just because both parts of the query (the estimation and the actual, original query) call into the `fetchChunks` method doesn't mean that both times is equally expensive. Maybe there's a bug somewhere and we actually fetch columns (as opposed to metadata) for the initial `RelAlgExecutor::getFilteredCountAll` call, in which case I want to know about it. But having some logs / timing information is not nearly enough to reach that conclusion -- it might be that reading the metadata itself is slow.

**Comment by fleapapa** [comment]
Let me probe into more details.

**Comment by fleapapa** [comment]
@asuhan. As you suggested:

`Yes, backup is a potential application for it. OTOH, backup can (and probably should) be implemented as a separate feature, it doesn't need to be in the roundabout way of running a SELECT * query.`

A direct path to get big SELECT result to the CSV file with tight memory and without inventing new wheels is what i am pursuing. Honestly i'm not sure yet at all whether this approach is reasonable or feasible.

**Comment by asuhan** [comment]
We want to dramatically reduce the output memory for projection queries which return a lot of results -- it's useful in many other contexts. For query like yours on wide tables, we can achieve 100x through entirely reasonable means -- maybe a week of work. Going further into the low-memory territory, we need to unpin fragments and throttle the execution, as we've already discussed. Also reasonable, but a bit kludgy since we'll need an adaptive strategy to decide when to throttle -- for the typical workload we target we don't want to throttle, we want maximum parallelism. We truly want the first part (projection size reduction), not sure about the value of the second part.

**Comment by fleapapa** [comment]
Dramatical memory reduction plus high parallelism is definitely the best solution for all use cases. Throttled execution is only an expedient, which seems to me reasonable only when there is no better solution and only when parallelized writes can't speed up the output a CSV file much (in my guess it's mostly an I/O bound task).

**Comment by asuhan** [comment]
Suppose we have the following query: `SELECT * FROM table WHERE filter(a, b, c ... z) > 0`. If there's not enough memory to hold enough chunks of `a`, `b` ... `z` in order to saturate the CPU cores, throttling is the only way and it'll entail loss of parallelism. Cutting down the size of the output buffer might or might not be enough -- throttling can still be needed if we genuinely want a system capable to operate with very low memory usage. Probably a non-goal for now.

**Comment by fleapapa** [comment]
@asuhan. You're right but to sooth my curiosity i added a tentative callback in ExecutionOptions to smuggle query results on a fragment by fragment basis. With low memory i can have only one thread at a time; otherwise three buffer slabs (4GB each) would be quickly filled up.

That's not enough. I also need to free all chunk buffers, clear **ExecutionDispatch::all_fragment_results_** and free all resources of **row_set_mem_owner_** immediately after query result of a fragment is smuggled out to ExportQueryStmt node by the callback; otherwise i will run out of memory on 2nd or 3rd fragment. 

With these measure, a footprint of 4GB suffices to run to the last fragment.

However, i still have one issue unsolved. After i materialize the smuggled result sets to CSV file by **ResultSet::getNextRow** iterations, the space used by the materialized result sets is sticky and not released even after **all_fragment_results_** is cleared. (A quick pointer will be very helpful!)

COPY big table with low memory seems not a urgent item. Though considerable number of developers in the community may have similar memory constraint, they might practice on unimportant sample data so none seemed bother to backup data.

(Note: regarding freeing all chunks, BufferMgr::deleteBuffersWithPrefix seems not working with prefix -1 ?)

**Comment by fleapapa** [comment]
[UPDATE]: The sticky memory turns out to be held by some StringDictionaryProxy. Finding how to manually release that memory before the active row_set_mem_owner_ is destructed...

**Comment by asuhan** [comment]
Both the fragments and string dictionary memory are sticky for good reasons. The rows need to be iterated, so they need access to that memory. To minimize the amount of used memory, we'd have to store metadata instead of actual chunks and "revive" them one by one, from disk, as we iterate. Is your thesis that we somehow leak memory or keep it around past its usefulness in the context of the current approach?

**Comment by fleapapa** [comment]
@asuhan. No. I know the dictionary needs to be sticky for some purposes like you mentioned. However, to minimize memory use to make room for next fragment, my (smuggling/streaming) experiment needs the dictionary to be closed after current result set is materialized.
